### DISCUSSION TITLE
liquidsoap 2.3.0* unable to grab mpegts and avoid transcoding

### DISCUSSION THREAD

**Discussion by xogium** [savonet/liquidsoap#3430] [User Support]
I'm trying to use the following script in order to grab the audio and video from obs and pass it to liquidsoap. Since obs is already encoding, I'm attempting to use the audio and video copy of ffmpeg in order to avoid encoding again.

The problem comes from the input itself. Liquidsoap seems unable to find a decoder to handle the stream. It is a standard stream from obs, mpegts format, content_type=video/m2ts. The audio is mp2 and the video codec is mpeg2video. Nothing that ffmpeg should have trouble decoding, I imagine.

```
live = input.harbor("harbor",port=8006,user="username",password="password",icy=true,buffer=6.0,max=12.0)

output.icecast(%ffmpeg(format="mpegts",
		%audio.copy,
		%video.copy),
	send_icy_metadata = true,
	transport = http.transport.tls(),
	host = "radio.domain.tld",
	port = 8443,
	mount = "/vstream.ts",
	password = "password",
	live)
```

The error I get is ```2023/09/23 16:41:37 [decoder:3] Unable to find a decoder for stream mime-type "video/m2ts" with expected content {audio=ffmpeg.copy,video=ffmpeg.copy}!```

What am I missing/doing wrong?

**Comment by vitoyucepi** [comment]
Hi @xogium,

I need to ask a few questions so I can try to reproduce your problem

1. What OS are you running liquidsoap on?
1. What version of liquidsoap are you running?
1. How did you install liquidsoap?

> It is a standard stream from obs, mpegts format, content_type=video/m2ts. The audio is mp2 and the video codec is mpeg2video.

![Obs recording settings in advanced mode](https://github.com/savonet/liquidsoap/assets/35035879/9bc67acf-f13b-4f9e-8aab-ec84371b6eee)

Looks not very standard. Is it necessary to stream in `mpeg2` to icecast?

Why wouldn't you use `input.rtmp` or `input.srt`?

**Comment by xogium** [comment]
Hi!
I run liquidsoap installed via opam on archlinux. I currently run the very latest code in the main branch.

As for why I did icecast, well. I initially streamed directly to icecast using this technic. I noticed a few problems with it, such as the fact that you can't use multiple words when defining stream name and genre, no matter if you quote them or not. Another advantages of using ls is that I can mix in audio of the stream with a playlist in the background eventually -- obs is on windows, ls is on linux.

As for why no rtmp/srt, I simply have no clue how these works, nor how to configure them in obs.

Hope this helps!

**Comment by xogium** [comment]
One thing I forgot to mention is that I am unsure what format to use other than this. I've tried webm but it keeps crashing obs unfortunately, so I stick with the format that wors, which just so happened to be mpegts.

**Comment by vitoyucepi** [comment]
> As for why no rtmp/srt, I simply have no clue how these works, nor how to configure them in obs.

To stream from OBS you should configure *custom* server in the *stream* settings.
![OBS rtmp settings](https://github.com/savonet/liquidsoap/assets/35035879/13a1c07c-ad15-421f-a00c-247c6e1ae0dc)
And use the *streaming* tab in the output settings.
![OBS stream settings](https://github.com/savonet/liquidsoap/assets/35035879/bc9b1f27-3599-49bd-a5a5-558cfe0fcf3d)

The liquidsoap `live` could be configured like this
```ruby
live = input.rtmp('rtmp://0.0.0.0:8000/a/b')
```

> installed via opam on archlinux

Also available in [aur](https://aur.archlinux.org/packages/liquidsoap).

**Comment by xogium** [comment]
Hi,
unfortunately the screenshots you've sent are not useful to me since I'm blind. Can you explain in text please?

As for the aur package, I've tried to use that, but I've had too many issues with it and stick to opam these days. I kept having problems to compile this or that, and it was just too much mess to maintain. Opam is much easier. But thanks for pointing it out!

**Comment by vitoyucepi** [comment]
> Can you explain in text please?

Sure.
1. Open settings: *File* -> *Settings*.
1. In the left column navigate to *Stream*.
1. In the right pane from the *Service* dropdown list select *Custom*.
1. In the *Server* text input field write `rtmp://YOUR_SERVER_ADDRESS:PORT/ENDPOINT`.
   Replace `YOUR_SERVER_ADDRESS` with the correct address of your server.
   Replace `PORT` with the port of the server, for example `8000`.
   Replace `ENDPOINT` with the path to the application, for example, `app`.
1. In the *Stream* key input field write `YOUR_SECRET_KEY`, for example, `b`.
1. In the left column navigate to *Streaming*.
1. In the right pane from the *Output Mode* dropdown list select *Simple*.
1. In the *Video Bitrate* input field write the video bitrate, for example, *2500 Kbps*.
1. From the *Audio Bitrate* dropdown list select the audio bitrate, for example *160*.
1. From the *Video Encoder* dropdown list select the video encoder, for example, *Software (x264)*.
1. From the *Encoder Preset* dropdown list select the encoder speed/quality, for example, *veryfast*.
1. Don't enable *Custom Encoder Settings*.
1. From the *Audio Encoder* dropdown list select the audio encoder, for example *AAC*.
1. That's all, you can now save settings.

To configure the liquidsoap you can use `input.rtmp`.
This function accepts string in format `"rtmp://IP_ADDRESS:PORT/ENDPOINT/YOUR_SECRET_KEY"`
- To listen for incoming connections from anywhere replace the `IP_ADDRESS` with `0.0.0.0`.
- Replace `PORT` with any port, for example `8000`. Use the same port as in the OBS config.
- Replace `ENDPOINT` with the path to the application, for example, `app`. Use the same path as in the OBS config.
- Replace `YOUR_SECRET_KEY` with the key, for example, `b`. Use the same path as in the OBS config.

You can store encoder settings in the variable.
`enc = %ffmpeg(%audio.copy, %video.copy, format="mpegts")`

Here's the resulting liquidsoap script
```ruby
live = input.rtmp("rtmp://0.0.0.0:8000/app/b")

enc = %ffmpeg(%audio.copy, %video.copy, format="mpegts")

output.icecast(
  enc,
  live,
  transport = http.transport.tls(),
  host = "radio.domain.tld",
  port = 8443,
  mount = "/vstream.ts",
  password = "password",
)
```

But there's a catch ffmpeg has not implemented strict verification for endpoint and key.
The [patch](https://patchwork.ffmpeg.org/project/ffmpeg/patch/20190925185708.70924-1-unique.will.martin@gmail.com/) was discussed but hasn't been merged.

The other possibility is to use SRT.
OBS has a [guide](https://obsproject.com/kb/srt-protocol-streaming-guide) about SRT streaming. The header of the section is *Stream with SRT*.
To stream using SRT you should change the settings.
1. Open settings: *File* -> *Settings*.
1. In the left column navigate to *Stream*.
1. In the right pane from the *Service* dropdown list select *Custom*.
1. In the *Server* text input field write `srt://YOUR_SERVER_ADDRESS:PORT?passphrase=YOUR_PASSPHRASE`.
   Replace `YOUR_PASSPHRASE` with a random passphrase with a length from 10 to 79 symbols.

In the liquidsoap config use the `input.srt` function instead of `input.rtmp`.
The `input.srt` function has a lot of options.
The most useful are
- `bind_address="IP_ADDRESS"` - the server listen address, the same as in *rtmp*.
  The default value is `0.0.0.0`.
- `port=PORT` - the port of the *srt* server, similar to *rtmp*.
  The default value is `8000`.
- `passphrase="YOUR_PASSPHRASE"` - the secret password required to connect to the stream.
  Use the same value as in the OBS settings.
- `enforced_encryption=true` - used in combination with the *passphrase* to ensure authentication and authorization.

The liquidsoap with the SRT will look like
```ruby
live = input.srt(
  bind_address="0.0.0.0",
  port=8000,
  passphrase="YOUR_PASSPHASE",
  enforced_encryption=true,
)

enc = %ffmpeg(%audio.copy, %video.copy, format="mpegts")

output.icecast(
  enc,
  live,
  transport = http.transport.tls(),
  host = "radio.domain.tld",
  port = 8443,
  mount = "/vstream.ts",
  password = "password",
)
```

**Reply by rrolla** [reply]
@vitoyucepi Did you manage also to get working video+audio stream from obs to liquidsoap?

One thing i noticed i need to forward **udp** port on docker, otherwise connection never reach liquidsoap from obs and hls processing doesnt start

```ruby
settings.log.level.set(5)
settings.server.log.level.set(5)
settings.log.stdout.set(true)

settings.srt.log.set(true)
settings.srt.log.level.set(5)
settings.srt.log.verbosity.set("debug")
settings.ffmpeg.log.verbosity.set("verbose")

set("frame.video.width", 1920)
set("frame.video.height", 1080)
set("frame.video.samplerate", 25)

def onConnect()
    print("Live connected!")
    {}
end

liveSource = input.srt(bind_address="0.0.0.0", port=38112, on_connect=onConnect, enforced_encryption=false)

### Outputs

# Define outputs formats

flvSourceEncoder = %ffmpeg(
    format="flv",
    %audio.copy,
    %video(codec="libx264", pixel_format="yuv420p", b="3000k", preset="veryfast", g=25),
  ).{extname = ".ts"}

# Output to HLS

def segment_name(~position, ~extname, stream_name) =
    timestamp = int_of_float(time())
    duration = 2
    "#{stream_name}_#{duration}_#{timestamp}_#{position}#{extname}"
end
  
def on_file_change(~state,fname) =
    if state == "closed" and file.extension(fname) != '.m3u8' then
        log.important(label="hls", "Segment #{fname} created")
    end
end

streams = [("flvSource", flvSourceEncoder)]

output.file.hls(id="output_hls",
                fallible=true,
                playlist="live.m3u8",
                segment_duration=2.0,
                segments=5,
                segments_overhead=5,
                segment_name=segment_name,
                on_file_change=on_file_change,
                persist_at="state.config",
                temp_dir= "/var/liquidsoap/hls",
                "/var/liquidsoap/hls",
                streams,
                liveSource)
```


```bash
docker run --rm --name liquidsoap -p 38112:38112/udp -v $(pwd)/liquidsoap:/var/liquidsoap urdtlv/liquidsoap:2.2.5-alpine /var/liquidsoap/live.liq
```

On OBS Settings -> Stream
```
Service: Custom...
Server: srt://localhost:38112
```

<details>

<summary>Logs from liquidsoap</summary>


```bash
2024/05/26 14:10:58 [video.text:3] Using sdl implementation
2024/05/26 14:11:19 [srt:5] 14:11:19.109847/SRT:RcvQ:w1.N:SRT.cn: PASSING request from: 172.17.0.1:47476 to agent:894745397
2024/05/26 14:11:19 [srt:5] 14:11:19.109986/SRT:RcvQ:w1.N:SRT.cn: Listener managed the connection request from: 172.17.0.1:47476 result:waveahand
2024/05/26 14:11:19 [srt:5] 14:11:19.110093/SRT:RcvQ:w1.N:SRT.cn: PASSING request from: 172.17.0.1:47476 to agent:894745397
2024/05/26 14:11:19 [srt:5] 14:11:19.110117/SRT:RcvQ:w1 D:SRT.sm: generateSocketID: : @894745396
2024/05/26 14:11:19 [srt:5] 14:11:19.110210/SRT:RcvQ:w1.N:SRT.cn: HSREQ/rcv: cmd=1(HSREQ) len=12 vers=0x10404 opts=0xbf delay=120
2024/05/26 14:11:19 [srt:5] 14:11:19.110536/SRT:RcvQ:w1.N:SRT.cn: @894745397: listen ret: -1 - conclusion
2024/05/26 14:11:19 [srt:5] 14:11:19.110542/SRT:RcvQ:w1.N:SRT.cn: Listener managed the connection request from: 172.17.0.1:47476 result:waveahand
Live connected!
2024/05/26 14:11:19 [input.srt:4] New connection from 172.17.0.1:47476
2024/05/26 14:11:19 [decoder:4] Available decoders:
2024/05/26 14:11:19 [decoder:4] ffmpeg (priority: 10)
2024/05/26 14:11:19 [decoder:4] Selected decoder ffmpeg for mime-type application/ffmpeg with expected content {audio=ffmpeg.copy,video=canvas}
[h264 @ 0x79554bc6e000] Reinit context to 1920x1088, pix_fmt: yuv420p
[h264 @ 0x79554bc6e000] Increasing reorder buffer to 1
[buffer @ 0x79554bc6b480] w:1920 h:1080 pixfmt:yuv420p tb:1/90000 fr:0/1 sar:0/1
[buffer @ 0x79554bc6b480] TB:0.000011 FRAME_RATE:nan SAMPLE_RATE:nan
[fps @ 0x79554bc6b6c0] Set first pts to (in:0 out:0) from start time 0.000000
[fps @ 0x79554bc6b6c0] fps=25/1
[h264 @ 0x79554bc6f400] Reinit context to 1920x1088, pix_fmt: yuv420p
2024/05/26 14:11:20 [input.srt:5] Got metadata at position 0: calling handlers...
2024/05/26 14:11:20 [input.srt:5] Got metadata at position 0: calling handlers...
2024/05/26 14:11:20 [output_hls:4] Resuming from saved state
2024/05/26 14:11:20 [output_hls:5] Opening segment 129 for stream flvSource.
[buffer @ 0x79554bc6ba80] w:1920 h:1080 pixfmt:yuv420p tb:1/25 fr:0/1 sar:1/1
[buffer @ 0x79554bc6ba80] TB:0.040000 FRAME_RATE:nan SAMPLE_RATE:nan
[fps @ 0x79554bc6bc00] fps=25/1
Automatically inserted bitstream filter 'aac_adtstoasc'; args=''
2024/05/26 14:11:20 [clock.input.srt:3] Delegating synchronization to active sources
[fps @ 0x79554bc6bc00] Set first pts to 0
2024/05/26 14:11:25 [output_hls:4] Terminating current segment on stream flvSource to make expected length
2024/05/26 14:11:25 [output_hls:5] Cleaning up /var/liquidsoap/hls/flvSource_109..ts..
2024/05/26 14:11:25 [output_hls:5] Writing playlist flvSource..
2024/05/26 14:11:25 [output_hls:5] Writing playlist /var/liquidsoap/hls/live.m3u8..
2024/05/26 14:11:25 [output_hls:5] Opening segment 130 for stream flvSource.
2024/05/26 14:11:31 [output_hls:4] Terminating current segment on stream flvSource to make expected length
2024/05/26 14:11:31 [output_hls:5] Cleaning up /var/liquidsoap/hls/flvSource_110..ts..
2024/05/26 14:11:31 [output_hls:5] Writing playlist flvSource..
2024/05/26 14:11:31 [output_hls:5] Opening segment 131 for stream flvSource.
```

</details>

I can see that playlists files and segment **.ts** are generated, but when i try to open with vlc (ubuntu 22.04.4 v3.0.16 Vetinari) or hls.js (https://players.akamai.com/players/hlsjs) it doesn't start play

I can see also in nginx that server hls generated files all return 200

**Comment by xogium** [comment]
Thank you so much! This works as I need now.

Fwiw, I chose srt, seemed the less annoying of them to set up.

I'm still curious why the mpegts stream didn't want to get through ffmpeg, that said. Any ideas on that one?

**Comment by vitoyucepi** [comment]
> I'm still curious why the mpegts stream didn't want to get through ffmpeg, that said. Any ideas on that one?

You can enable option *Show all codecs*, like it's done in the [ePirat's blog post](https://epir.at/2018/03/08/obs-icecast-streaming/).
This option will enable you to choose *libx264* as the video encoder and *aac* as the audio encoder.
But there was an error. The configured value for *Muxer Settings* is `content_type=video/mp2ts`. This is the incorrect mime type for mpegts. It should be *video/mp2t*, not *video/mp2ts*. It's also possible to leave this field empty.

1. In the left column navigate to *Output*.
1. In the right pane from the *Output Mode* dropdown list select *Advanced*.
1. Navigate to *Recording*.
1. From the *Type* dropdown list select *Custom Output (FFmpeg)*.
1. From the *FFmpeg Output Type* dropdown list select *Output to URL*.
1. In the *File path or URL* text input field write the Icecast URL, for example, `icecast://source:hackme@127.0.0.1:8000/example.ts`.
1. From the *Container Format* dropdown list select *mpegts*.
1. In the *Muxer Settings* text input field write muxer settings. You can leave this field empty.
1. In the *Video Bitrate* number input field write the desired video bitrate, for example, *2500 Kbps*.
1. In the *Keyframe interval* number input field write the interval between the key frames. The recommended number for streaming is 2 seconds multiplied by FPS. For 30 frames per second, this value should be 60.
1. Enable option *Show all codecs*.
1. From the *Video Encoder* dropdown list select a video codec, for example, *libx264*.
1. In the *Video Encoder Settings* text input field write the settings like preset.
1. In the *Audio Bitrate* number input field write the desired audio bitrate, for example, *160 Kbps*.
1. From the *Audio Encoder* dropdown list select an audio codec, for example *aac*.

By the way, I have found another type of error. It's not possible to set another encoder without restarting the liquidsoap when copy-codec is used.

**Comment by toots** [comment]
Hi all! This great and helpful. For future reference, decoder mime types for `ffmpeg` (and others) can be set at runtime using `settings.decoder.mime_types.ffmpeg`: [doc here](https://www.liquidsoap.info/doc-2.2.0/settings.html#mime-types-used-for-decoding-with-ffmpeg)

`input.harbor` wouldn't seem like the most suited input for OBS indeed. The two typical method with it are `rtmp` or `srt`.

Thanks for all the details. I'm converting this to a discussion so it can stick around.
